File: LeetCode/leetCode-Valid_Parentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function (s) {
  const stack = [];
  for (let i = 0; i < s.length; i++) {
    if (s[i] === '(' || s[i] == '[' || s[i] == '{') {
      stack.push(s[i]);
    } else if (s[i] === ')') {
      if (s[i - 1] == '(') {
        if (stack.length !== 0) {
          stack.pop();
        } else {
          return false;
        }
      } else {
        if (stack.length !== 0) {
          if (stack.pop() === '(') {
            continue;
          } else {
            return false;
          }
        } else {
          return false;
        }
      }
    } else if (s[i] === ']') {
      if (s[i - 1] == '[') {
        if (stack.length !== 0) {
          stack.pop();
        } else {
          return false;
        }
      } else {
        if (stack.length !== 0) {
          if (stack.pop() === '[') {
            continue;
          } else {
            return false;
          }
        } else {
          return false;
        }
      }
    } else if (s[i] === '}') {
      if (s[i - 1] == '{') {
        if (stack.length !== 0) {
          stack.pop();
        } else {
          return false;
        }
      } else {
        if (stack.length !== 0) {
          if (stack.pop() === '{') {
            continue;
          } else {
            return false;
          }
        } else {
          return false;
        }
      }
    }
  }
  if (stack.length !== 0) {
    return false;
  } else {
    return true;
  }
};

File: LeetCode/leetCode-Valid-Parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        s = list(s)
        stack = []
        for i in range(len(s)) :
            if s[i] == '(' or s[i] == '[' or s[i] == '{' :
                stack.append(s[i])
            elif s[i] == ')' :
                if s[i-1] == '(' :
                    if len(stack) != 0 :
                        stack.pop()
                    else :
                        return False
                else :
                    if len(stack) != 0 :
                        if stack.pop() == '(' :
                            continue
                        else :
                            return False
                    else :
                        return False
            elif s[i] == ']' :
                if s[i-1] == '[' :
                    if len(stack) != 0 :
                        stack.pop()
                    else :
                        return False
                else :
                    if len(stack) != 0 :
                        if stack.pop() == '[' :
                            continue
                        else :
                            return False
                    else :
                        return False
            elif s[i] == '}' :
                if s[i-1] == '{' :
                    if len(stack) != 0 :
                        stack.pop()
                    else :
                        return False
                else :
                    if len(stack) != 0 :
                        if stack.pop() == '{' :
                            continue
                        else :
                            return False
                    else :
                        return False

        if len(stack) != 0 :
            return False
        else :
            return True

        